kratos/containers/container_expression/expressions/expression_iterator.h

```diff
@@ -0,0 +1,283 @@
+#include "containers/container_expression/expressions/literal/literal_flat_expression.h"
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @brief Iterator class for expressions generic lazy expressions
+ *
+ */
+template<class TExpressionType>
+class ExpressionIterator {
+public:
+    ///@name Type definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    using DataType = double;
+
+    using ConstIteratorType = ExpressionIterator<Expression>;
+
+    KRATOS_CLASS_POINTER_DEFINITION(ExpressionIterator);
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    ExpressionIterator(Expression::Pointer pExpression)
+        : mpExpression(pExpression),
+          mEntityIndex(0),
+          mEntityDataBeginIndex(0),
+          mComponentIndex(0),
+          mFlattenedShapeSize(pExpression->GetFlattenedShapeSize())
+    {}
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    Expression::Pointer GetExpression() { return mpExpression; }
+
+    ConstIteratorType cbegin() { return ConstIteratorType(mpExpression); }
+
+    ConstIteratorType cend()
+    {
+        ConstIteratorType result(mpExpression);
+        result.mEntityIndex = mpExpression->NumberOfEntities();
+        result.mEntityDataBeginIndex = result.mEntityIndex * result.mFlattenedShapeSize;
+        return result;
+    }
+
+    ///@}
+    ///@name Public operators
+    ///@{
+
+    double operator*() const
+    {
+        return mpExpression->Evaluate(mEntityIndex, mEntityDataBeginIndex, mComponentIndex);
+    }
+
+    bool operator==(const ConstIteratorType& rOther) const
+    {
+        return (&*mpExpression == &*rOther.mpExpression) && (mEntityIndex == rOther.mEntityIndex) && (mComponentIndex == rOther.mComponentIndex);
+    }
+
+    bool operator!=(const ConstIteratorType& rOther) const
+    {
+        return !this->operator==(rOther);
+    }
+
+    ConstIteratorType operator+(const IndexType n) const
+    {
+        ConstIteratorType result(mpExpression);
+        result.mComponentIndex = (mComponentIndex + n) % mFlattenedShapeSize;
+        result.mEntityIndex += n / mFlattenedShapeSize;
+        result.mEntityDataBeginIndex = result.mEntityIndex * mFlattenedShapeSize;
+        return result;
+    }
+
+    ConstIteratorType& operator++()
+    {
+        ++mComponentIndex;
+        if (mComponentIndex == mFlattenedShapeSize) {
+            mComponentIndex = 0;
+            ++mEntityIndex;
+            mEntityDataBeginIndex = mEntityIndex * mFlattenedShapeSize;
+        }
+        return *this;
+    }
+
+    ConstIteratorType operator++(int)
+    {
+        ConstIteratorType temp = *this;
+        ++*this;
+        return temp;
+    }
+
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    Expression::Pointer mpExpression;
+
+    IndexType mEntityIndex;
+
+    IndexType mEntityDataBeginIndex;
+
+    IndexType mComponentIndex;
+
+    IndexType mFlattenedShapeSize;
```

Comment:
I can't say I'm comfortable with this representation. As far as I understand, this iterator represents a scalar component within a 1D array holding flattened entities (entities can be scalars, vectors or matrices). I assume the expression can be queried for the flattened size of a single entity. If so, `mEntityIndex`, `mEntityDataBeginIndex`, and `mComponentIndex` can all be deduced from an absolute index within the 1D array, so they shouldn't be separately stored and managed by the iterator.

The scary downside of keeping track of all these separate indices is that implementing `operator-`, `operator<` and the like, which are required for random access iterators, becomes incredibly tedious and prone to bugs.

I recommend storing only the pointer to the expression and the absolute index of the component within the 1D array. Comparisons and increments/decrements/steps could be performed directly on the absolute index. Other indices could then be computed only when necessary (eg.: ˙operator*`).